buildSrc/src/main/java/net/neoforged/neodev/installer/LibraryCollector.java

```diff
@@ -0,0 +1,174 @@
+package net.neoforged.neodev.installer;
+
+import net.neoforged.neodev.utils.MavenIdentifier;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HexFormat;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import java.util.function.Function;
+
+/**
+ * For each file in a collection, finds the repository that the file came from.
+ */
+class LibraryCollector {
+    public static List<Library> resolveLibraries(List<URI> repositoryUrls, Collection<IdentifiedFile> libraries) throws IOException {
+        var collector = new LibraryCollector(repositoryUrls);
+        for (var library : libraries) {
+            collector.addLibrary(library.getFile().getAsFile().get(), library.getIdentifier().get());
+        }
+
+        var result = collector.libraries.stream().map(future -> {
+            try {
+                return future.get();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }).toList();
+        LOGGER.info("Collected %d libraries".formatted(result.size()));
+        return result;
+    }
+
+    private static final Logger LOGGER = Logging.getLogger(LibraryCollector.class);
+    /**
+     * Hosts from which we allow the installer to download.
+     * We whitelist here to avoid redirecting player download traffic to anyone not affiliated with Mojang or us.
+     */
+    private static final List<String> HOST_WHITELIST = List.of(
+            "minecraft.net",
+            "neoforged.net",
+            "mojang.com"
+    );
+
+    private static final URI MOJANG_MAVEN = URI.create("https://libraries.minecraft.net");
+    private static final URI NEOFORGED_MAVEN = URI.create("https://maven.neoforged.net/releases");
+
+    private final List<URI> repositoryUrls;
+
+    private final List<Future<Library>> libraries = new ArrayList<>();
+
+    private final HttpClient httpClient = HttpClient.newBuilder().build();
```

Comment:
The default http client uses normal threads, I wonder if we gain a speed benefit by using virtual threads.

Reply:
No idea, this is mostly adapted from NG.